File: Problem Solving/C++/ValidParenthesis.cpp
#include <iostream>
#include <stack>

using namespace std;

    bool matching(char a, char b) {
        if (a == '(' && b == ')' || a == '[' && b == ']' || a == '{' && b == '}') return true;
        return false;
    }
    bool isValid(string s) {
        stack<char> st;
        for (int i = 0 ; i < s.length() ; i++) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') st.push(s[i]);
            else {
                if (!st.empty()) {
                    if (!matching(st.top(), s[i])) return false;
                    else st.pop();
                }
                else st.push(s[i]);
            }
        }
        if (st.empty()) return true;
        else return false;
    } 

int main() {
    string s = "(){}]";
    cout << isValid(s);
}

File: Problem Solving/Java/ValidParenthesis.java
import java.util.Stack;

public class ValidParenthesis {
    public boolean matching(char a, char b) {
        if (a == '[' && b == ']' || a == '(' && b == ')' || a == '{' && b == '}') return true;
        else return false;
    }

    public boolean isValid(String s) {
        Stack<Character> st = new Stack<Character>();
        for (int i = 0 ; i < s.length() ; i++) {
            if (s.charAt(i) == '[' || s.charAt(i) == '(' || s.charAt(i) == '{') {
                st.push(s.charAt(i));
            }
            else {
                if (!st.empty()) {
                    if (!matching(st.peek(), s.charAt(i))) {
                        return false;
                    }
                    else {
                        st.pop();
                    }
                }
                else {
                    st.push(s.charAt(i));
                }
            }
        }
        if (st.empty()) return true;
        else return false;       
    }    
}